pallets/phala-world/src/incubation.rs

```diff
@@ -153,11 +245,123 @@ pub mod pallet {
 		pub fn hatch_origin_of_shell(
 			origin: OriginFor<T>,
+			owner: T::AccountId,
 			collection_id: CollectionId,
 			nft_id: NftId,
+			resource_src: BoundedVec<u8, <T as pallet_uniques::Config>::StringLimit>,
 		) -> DispatchResult {
-			let sender = ensure_signed(origin)?;
-			// TODO: Move to incubation.rs
+			let sender = ensure_signed(origin.clone())?;
+			pallet_pw_nft_sale::pallet::Pallet::<T>::ensure_overlord(sender.clone())?;
+			// Ensure that the collection is an Origin of Shell Collection
+			ensure!(
+				Self::is_origin_of_shell_collection_id(collection_id),
+				Error::<T>::WrongCollectionId
+			);
+			// Check if HatchTimes is less than or equal to current Timestamp
+			// Ensure that Origin of Shell exists or is not past the hatch time
+			let hatch_time = Self::get_hatch_time(collection_id, nft_id);
+			ensure!(Self::can_hatch(hatch_time), Error::<T>::CannotHatchOriginOfShell);
+			// Check if Shell Collection ID is set
+			let shell_collection_id = Self::get_shell_collection_id()?;
+			// Get race, key and origin of shell type before burning origin of shell NFT
+			let race_key = pallet_pw_nft_sale::pallet::Pallet::<T>::to_boundedvec_key("race")?;
+			let career_key = pallet_pw_nft_sale::pallet::Pallet::<T>::to_boundedvec_key("career")?;
+			let origin_of_shell_type_key =
+				pallet_pw_nft_sale::Pallet::<T>::to_boundedvec_key("origin_of_shell_type")?;
+			let race = pallet_uniques::Pallet::<T>::attribute(&collection_id, &nft_id, &race_key)
+				.ok_or(Error::<T>::RaceNotDetected)?;
+			let career =
+				pallet_uniques::Pallet::<T>::attribute(&collection_id, &nft_id, &career_key)
+					.ok_or(Error::<T>::CareerNotDetected)?;
+			let origin_of_shell_type_value = pallet_uniques::Pallet::<T>::attribute(
+				&collection_id,
+				&nft_id,
+				&origin_of_shell_type_key,
+			)
+			.ok_or(Error::<T>::OriginOfShellTypeNotDetected)?;
+			let race_type: RaceType = RaceType::from_u8(race[0]).expect("[race] should not fail");
+			let career_type: CareerType =
+				CareerType::from_u8(career[0]).expect("[career] should not fail");
+			let origin_of_shell_type: OriginOfShellType =
+				OriginOfShellType::from_u8(origin_of_shell_type_value[0])
+					.expect("[origin_of_shell_type] should not fail");
+			// Get Shell Collection next NFT ID
+			let shell_nft_id = pallet_rmrk_core::NextNftId::<T>::get(shell_collection_id);
+			// Get empty metadata
+			let metadata = pallet_pw_nft_sale::pallet::Pallet::<T>::get_empty_metadata();
+			// Get resource id
+			let resource_id = pallet_rmrk_core::Pallet::<T>::get_next_resource_id()?;
+			let resource_id: BoundedResource<T::ResourceSymbolLimit> =
+				resource_id.encode().try_into().expect("[resource] id should work");
+			// Burn Origin of Shell NFT then Mint Shell NFT
+			pallet_rmrk_core::Pallet::<T>::burn_nft(
+				Origin::<T>::Signed(owner.clone()).into(),
+				collection_id,
+				nft_id,
+			)?;
+			// Remove Attributes from Uniques pallet
+			pallet_uniques::Pallet::<T>::clear_attribute(
```

Comment:
Aren't the attributes got clear after the burning?

Reply:
From parity, it seems that downstream pallets must manage the attributes. I could create another issue to ask for this to be implemented.

https://github.com/paritytech/substrate/issues/9207#issuecomment-890367408